runelite-client/src/main/java/net/runelite/client/plugins/zeahrunecraft/ZeahRunecraftPlugin.java

```diff
@@ -0,0 +1,170 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.zeahrunecraft;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import com.google.inject.Provides;
+
+import lombok.Getter;
+import net.runelite.api.Client;
+import net.runelite.api.GameObject;
+import net.runelite.api.Player;
+import net.runelite.api.events.AnimationChanged;
+import net.runelite.api.events.GameObjectDespawned;
+import net.runelite.api.events.GameObjectSpawned;
+import net.runelite.api.events.GameStateChanged;
+import net.runelite.api.events.VarbitChanged;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.eventbus.Subscribe;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import net.runelite.client.ui.overlay.OverlayManager;
+
+@PluginDescriptor(
+		name = "Zeah Runecraft",
+		description = "Show overlays when Runecrafting in Zeah",
+		tags = { "highlight", "overlay" },
+		enabledByDefault = true)
+public class ZeahRunecraftPlugin extends Plugin
+{
+	@Getter
+	private final List<DenseRunestone> runestones = new ArrayList<>();
+
+	@Getter
+	private boolean mining = false;
+
+	@Inject
+	private Client client;
+
+	@Inject
+	private OverlayManager overlayManager;
+
+	@Inject
+	private ZeahRunecraftOverlay piniOverlay;
+
+	@Provides
+	ZeahRunecraftConfig getConfig(final ConfigManager configManager)
+	{
+		return configManager.getConfig(ZeahRunecraftConfig.class);
+	}
+
+	@Override
+	protected void startUp() throws Exception
+	{
+		overlayManager.add(piniOverlay);
+	}
+
+	@Override
+	protected void shutDown() throws Exception
+	{
+		overlayManager.remove(piniOverlay);
+		runestones.clear();
+	}
+
+	@Subscribe
+	public void onGameStateChanged(final GameStateChanged event)
+	{
+		switch (event.getGameState())
+		{
+			case HOPPING:
+			case LOGIN_SCREEN:
+			case LOADING:
+				runestones.clear();
+				break;
+			default:
+				break;
+		}
+	}
+
+	@Subscribe
+	public void onAnimationChanged(final AnimationChanged event)
+	{
+		Player local = client.getLocalPlayer();
+
+		if (event.getActor() != local)
+		{
+			return;
+		}
+
+		int playerAnimationId = local.getAnimation();
+
+		mining = MiningAnimations.isPlayerMining(playerAnimationId);
+	}
+
+	@Subscribe
+	public void onGameObjectSpawned(final GameObjectSpawned event)
+	{
+		GameObject gameObject = event.getGameObject();
+		int gameObjectId = gameObject.getId();
+
+		// Check to see if the GameObject is a Dense Runestone

```

Comment:
All of this just reduces to `if (gameObjectId == obj1 || gameObjectId == obj2) runestones.add(...`

Reply:
Look at using `ObjectId` and/or `NullObjectId` for those constants. Are you sure duplicate spawn events happen? That shouldn't happen if you are handling both despawns and the game state change event. I think.